modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/MergeSortDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,298 @@
+            NodePageStream s = streams[i];
+
+            // Skip if stream has head: on previous next() head extracts from stream[streamOff], others streams do
+            // have head or it is the first run.
+            if (s.head() != null)
+                break;
+
+            if (!s.hasNext()) {
+                // Nullify obsolete stream, move left bound.
+                streamsMap.remove(s.nodeId);
+                streams[i] = null;
+                streamOff++;
+            }
+            else {
+                // Prefetch head value.
+                s.next();
+            }
+        }
+
+        if (finished())
+            throw new NoSuchElementException("No next element. Please, be sure to invoke hasNext() before next().");
+
+        if (first) {
+            first = false;
+
+            Arrays.sort(streams, streamCmp);
+        } else
+            bubbleUp(streams, streamOff, streamCmp);
+
+        return streams[streamOff].get();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() throws IgniteCheckedException {
+        for (int i = streamOff; i < streams.length; i++) {
+            if (streams[i].hasNext())
+                return true;
+
+            // Nullify obsolete stream, move left bound.
+            streamsMap.remove(streams[i].nodeId);
+            streams[i] = null;
+            streamOff++;
+        }
+
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void loadPage() {
+        assert !Thread.holdsLock(queueLock());
+
+        List<UUID> nodes;
+
+        synchronized (queueLock()) {
+            if (!loadAllowed)
+                return;
+
+            nodes = new ArrayList<>();
+
+            for (int i = streamOff; i < streams.length; i++) {
+                UUID nodeId = streams[i].nodeId;
+
+                // Try to contain 2 pages for every stream to avoid waits. A node has to be present in collections:
+                // 1. rcvd - as requested and already handled before, so currently no parallel request to this node.
+                // 2. subgrid - as this node still has pages to request.
+                if (streams[i].queue.size() < 1 && rcvd.remove(nodeId) && subgrid.contains(nodeId))
+                    nodes.add(nodeId);
+            }
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onPage(@Nullable UUID nodeId, boolean last) {
+        boolean qryLast = super.onPage(nodeId, last);
+
+        if (nodeId == null)
+            nodeId = cctx.localNodeId();
+
+        if (last && streamsMap.containsKey(nodeId)) {
+            streamsMap.get(nodeId).allPagesReady = true;
+
+            streamsMap.remove(nodeId);
+        }
+
+        return qryLast;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addPage(@Nullable UUID nodeId, Collection<R> data) {
+        assert Thread.holdsLock(queueLock());
```

Comment:
Why do we care about external synchronization?
Seems,  it is enough to avoid NPE after streamsMap.get(nodeId).

Reply:
You're right. We need to check streamsMap.get() for NPE in case of concurrent `onProcessQueryResponse()` and `onNodeLeft()`. Guarding with lock doesn't help us in this case.

Reply:
You're right. I've simplified code, as nodeId == null for distributed reducer means only error. So we can remove check for sync (that was required to guard streamsMap) and replace with check for NPE.